File: Bronze/2017/December/02_the_bovine_shuffle/TheBovineShuffle.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.Scanner;
import java.util.TreeMap;

/**
 * @author Red
 * @version 1.0
 */
public class TheBovineShuffle {
    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("shuffle.in"));
        PrintWriter pw = new PrintWriter("shuffle.out");
        int n = Integer.parseInt(r.readLine());
        // shuffle 方式   当前顺序
        String[] shuffle = r.readLine().split(" ");
        String[] order = r.readLine().split(" ");
        // 根据 shuffle 方式重新排列三次
        for (int i = 0; i < 3; i++) {
            String[] temp = new String[n];
            for (int j = 0; j < n; j++) {
                temp[j] = order[Integer.parseInt(shuffle[j]) - 1];
            }
            order = temp;
        }
        for (int i = 0; i < n; i++) {
            pw.println(order[i]);
        }
        pw.close();
    }
}

File: Bronze/2017/December/02_the_bovine_shuffle/the_bovine_shuffle.cpp
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

int main() {
    freopen("shuffle.in", "r", stdin);
    freopen("shuffle.out", "w", stdout);
    int N;
    cin >> N;
    // shuffle 方式
    vector<int> shuffle(N);
    //  当前顺序
    vector<int> order(N);
    for (int i = 0; i < N; i++) {
        cin >> shuffle[i];
    }
    for (int i = 0; i < N; i++) {
        cin >> order[i];
    }
    // 根据 shuffle 方式重新排列三次
    for (int i = 0; i < 3; i++) {
        vector<int> temp(N);
        for (int j = 0; j < N; j++) {
            temp[j] = order[shuffle[j] - 1];
        }
        order = temp;
    }
    for (int i = 0; i < N; i++) {
        cout << order[i] << endl;
    }
    return 0;
}
